docs(stepfunctions-tasks): fill in README (#7469)
aws/aws-cdk: packages/@aws-cdk/aws-stepfunctions-tasks/README.md

---
<!--END STABILITY BANNER-->

[AWS Step Functions](https://docs.aws.amazon.com/step-functions/latest/dg/welcome.html) is a web service that enables you to coordinate the
components of distributed applications and microservices using visual workflows.
You build applications from individual components that each perform a discrete
function, or task, allowing you to scale and change applications quickly.

A [Task](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-task-state.html) state represents a single unit of work performed by a state machine. All work in your state
machine is performed by tasks.

This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project.

## Table Of Contents

> - [Task](#task)
>   - [Parameters](#task-parameters-from-the-state-json)
>   - [Batch](#batch)
>     - [SubmitJob](#submitjob)
>   - [DynamoDB](#dynamodb)
>     - [GetItem](#getitem)
>     - [PutItem](#putitem)
>     - [DeleteItem](#deleteitem)
>     - [UpdateItem](#updateitem)
>   - [ECS](#ecs)
>     - [RunTask](#runtask)
>   - [EMR](#emr)
>     - [Create Cluster](#create-cluster)
>     - [Termination Protection](#termination-protection)
>     - [Terminate Cluster](#terminate-cluster)
>     - [Add Step](#add-step)
>     - [Cancel Step](#cancel-step)
>     - [Modify Instance Fleet](#modify-instance-fleet)
>     - [Modify Instance Group](#modify-instance-group)
>   - [Evaluate Expression](#evaluate-expression)
>   - [Glue](#glue)
>   - [Lambda](#lambda)
>   - [SageMaker](#sagemaker)
>     - [Create Training Job](#create-training-job)
>     - [Create Transform Job](#create-transform-job)
>   - [SNS](#sns)
>   - [Step Functions](#step-functions)
>   - [SQS](#sqs)

Reviewer: I would lose one level of depth there. Something like this -

```suggestion
- [Task](#task)
- [Parameters](#task-parameters-from-the-state-json)
- [Batch](#batch)
  - [SubmitJob](#submitjob)
- [DynamoDB](#dynamodb)
  - [GetItem](#getitem)
  - [PutItem](#putitem)
  - [DeleteItem](#deleteitem)
  - [UpdateItem](#updateitem)
- [ECS](#ecs)
  - [RunTask](#runtask)
- [EMR](#emr)
  - [Create Cluster](#create-cluster)
  - [Termination Protection](#termination-protection)
  - [Terminate Cluster](#terminate-cluster)
  - [Add Step](#add-step)
  - [Cancel Step](#cancel-step)
  - [Modify Instance Fleet](#modify-instance-fleet)
  - [Modify Instance Group](#modify-instance-group)
- [Evaluate Expression](#evaluate-expression)
- [Glue](#glue)
- [Lambda](#lambda)
- [SageMaker](#sagemaker)
  - [Create Training Job](#create-training-job)
  - [Create Transform Job](#create-transform-job)
- [SNS](#sns)
- [Step Functions](#step-functions)
- [SQS](#sqs)
```

> ### Task

Reviewer: ```suggestion
## Task
```

> A `Task` represents some work that needs to be done. In the CDK, the exact work to be

Reviewer: I would use the exact same definition from the step functions docs - https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-task-state.html

```suggestion
A Task state represents a single unit of work performed by a state machine. In the CDK, the exact work to be
```

done is determine by a class that implements `IStepFunctionsTask`.

> AWS Step Functions [integrates](https://docs.aws.amazon.com/step-functions/latest/dg/concepts-service-integrations.html) with some AWS services so that you can call API

Reviewer: Use line width 120. Few other places in the doc as well.

Author: fixed up a few instances where line widths were inconsistent. I think I used an even smaller line width (100), but the ones with the links that extend beyond it. They're better when rendered.

actions, and coordinate executions directly from the Amazon States Language in
Step Functions. You can directly call and pass parameters to the APIs of those
services.

> #### Task parameters from the state JSON

Reviewer: ```suggestion
## Task parameters from the state JSON
```

> Many tasks take parameters. Parameter values can either be supplied
> directly in the workflow definition (by specifying their values), or at
> runtime by passing a value obtained from the static functions on `Data`,
> such as `Data.stringAt()`.

Reviewer: ```suggestion
Most tasks take parameters. Parameter values can either be static, supplied
directly in the workflow definition (by specifying their values), or a value
available at runtime in the state machine's execution (either as its input or
an output of a prior state).

Parameter values available at runtime can be specified via the `Data` class,
using methods such as `Data.stringAt()`.
```

If so, the value is taken from the indicated location in the state JSON,

> similar to (for example) `inputPath`.

Reviewer: A short code snippet may be useful here before this statement.

Author: I dropped this statement since it's illustrated across examples. Might be worth supplementing this section with a few examples of `path` usage which is what this really tries to touch on.

> #### Batch
> 
> Step Functions supports [Batch](https://docs.aws.amazon.com/step-functions/latest/dg/connect-batch.html) through the service integration pattern.
> 
> #### SubmitJob

Reviewer: I wouldn't go as deep as h4 and h5. How about keeping it at h2 and h3 instead?

```suggestion
## Batch

Step Functions supports [Batch](https://docs.aws.amazon.com/step-functions/latest/dg/connect-batch.html) through the service integration pattern.

### SubmitJob
```

Apply the same for all future sections.

```ts
new sfn.Task(stack, 'Task', {
  task: new tasks.EmrCancelStep({
    clusterId: 'ClusterId',
    stepId: 'StepId',
  }),
});
```

##### Modify Instance Fleet

Modifies the target On-Demand and target Spot capacities for the instance
fleet with the specified InstanceFleetName.

Corresponds to the [`modifyInstanceFleet`](https://docs.aws.amazon.com/emr/latest/APIReference/API_ModifyInstanceFleet.html) API in EMR.

```ts
new sfn.Task(stack, 'Task', {
  task: new tasks.EmrModifyInstanceFleetByName({
    clusterId: 'ClusterId',
    instanceFleetName: 'InstanceFleetName',
    targetOnDemandCapacity: 2,
    targetSpotCapacity: 0,
  }),
});
```

##### Modify Instance Group

Modifies the number of nodes and configuration settings of an instance group.

Corresponds to the [`modifyInstanceGroups`](https://docs.aws.amazon.com/emr/latest/APIReference/API_ModifyInstanceGroups.html) API in EMR.

```ts
new sfn.Task(stack, 'Task', {
  task: new tasks.EmrModifyInstanceGroupByName({
    clusterId: 'ClusterId',
    instanceGroupName: sfn.Data.stringAt('$.InstanceGroupName'),
    instanceGroup: {
      instanceCount: 1,
    },
  }),
});
```

> #### Evaluate Expression

Reviewer: This is a useful one, and not specific to any service integration. 
Wonder if we should bump this to the top of the list.

Use the `EvaluateExpression` to perform simple operations referencing state paths. The
`expression` referenced in the task will be evaluated in a Lambda function

> (`eval()`). This allows you to not have to write Lambda code for simple operations.

Reviewer: Provide a simple example of this maybe, along with the runtime prop?

Author: it's implied in the previous example since there's only one supported runtime. I amended the example to explicitly specify the runtime.
